test/blackbox/common/common_handlers.c

```diff
@@ -0,0 +1,250 @@
+/*
+    common_handlers.c -- Implementation of common callback handling and signal handling
+                            functions for black box tests
+    Copyright (C) 2017  Guus Sliepen <guus@meshlink.io>
+                        Manav Kumar Mehta <manavkumarm@yahoo.com>
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+#include <stdio.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/ioctl.h>
+#include <net/if.h>
+#include <arpa/inet.h>
+#include "test_step.h"
+#include "common_handlers.h"
+
+char *lxc_bridge = NULL;
+black_box_state_t *state_ptr = NULL;
+
+bool meta_conn_status[10];
+bool node_reachable_status[10];
+
+void mesh_close_signal_handler(int a) {
+    execute_close();
```

Comment:
Can you brief this a little bit more about this

Reply:
Yes: only a subset of function calls is safe to perform while inside a signal handler, see: http://man7.org/linux/man-pages/man7/signal-safety.7.html. The function `execute_close()` calls `meshlink_close()`, which calls many functions that are not in the list of safe functions.

In many test cases, I see the pattern: `setup_signals(); ...; execute_start(); while(1) select(...);`. So instead of calling `execute_close()` inside the signal handler, just set a global variable, say `test_running`, to `true`, and immediately return. Then inside `main()`, replace the while loop with: `while(test_running) pause(); execute_close();`.